bin/kmstool-enclave-cli/main.c

```diff
@@ -0,0 +1,218 @@
+    struct aws_nitro_enclaves_kms_client_configuration configuration = {
+        .allocator = app_ctx->allocator,
+        .endpoint = &endpoint,
+        .domain = AWS_SOCKET_VSOCK,
+        .region = app_ctx->region
+    };
+
+    /* SetCredentials operation sets the AWS credentials and creates a KMS
+        * client.with them. This needs to be called before Decrypt. */
+    struct aws_credentials *new_credentials = aws_credentials_new(
+        app_ctx->allocator,
+        aws_byte_cursor_from_c_str((const char*) app_ctx->aws_access_key_id->bytes),
+        aws_byte_cursor_from_c_str((const char*) app_ctx->aws_secret_access_key->bytes),
+        aws_byte_cursor_from_c_str((const char*) app_ctx->aws_session_token->bytes),
+        UINT64_MAX);
+
+    /* If credentials or client already exists, replace them. */
+    if (credentials != NULL) {
+        aws_nitro_enclaves_kms_client_destroy(client);
+        aws_credentials_release(credentials);
+    }
+
+    credentials = new_credentials;
+    configuration.credentials = new_credentials;
+    client = aws_nitro_enclaves_kms_client_new(&configuration);
+    
+    /* Decrypt uses KMS to decrypt the data passed to it in the Ciphertext
+    * field and sends it back to the called*
+    * TODO: This should instead send a hash of the data instead.
+    */
+
+    /* Get decode base64 string into bytes. */
+    size_t ciphertext_len;
+    struct aws_byte_buf ciphertext;
+    struct aws_byte_cursor ciphertext_b64 = aws_byte_cursor_from_c_str((const char*) app_ctx->ciphertext_b64->bytes);
+
+    rc = aws_base64_compute_decoded_len(&ciphertext_b64, &ciphertext_len);
+    fail_on(rc != AWS_OP_SUCCESS);
+    rc = aws_byte_buf_init(&ciphertext, app_ctx->allocator, ciphertext_len);
+    fail_on(rc != AWS_OP_SUCCESS);
+    rc = aws_base64_decode(&ciphertext_b64, &ciphertext);
+    fail_on(rc != AWS_OP_SUCCESS);
+
+    /* Decrypt the data with KMS. */
+    struct aws_byte_buf ciphertext_decrypted;
+    rc = aws_kms_decrypt_blocking(client, &ciphertext, &ciphertext_decrypted);
+    aws_byte_buf_clean_up(&ciphertext);
+    fail_on(rc != AWS_OP_SUCCESS);
+
+    /* Encode ciphertext into base64 for sending back result. */
+    size_t ciphertext_decrypted_b64_len;
+    struct aws_byte_cursor ciphertext_decrypted_cursor = aws_byte_cursor_from_buf(&ciphertext_decrypted);
+    aws_base64_compute_encoded_len(ciphertext_decrypted.len, &ciphertext_decrypted_b64_len);
+    rc = aws_byte_buf_init(ciphertext_decrypted_b64, app_ctx->allocator, ciphertext_decrypted_b64_len + 1);
+    fail_on(rc != AWS_OP_SUCCESS);
+    rc = aws_base64_encode(&ciphertext_decrypted_cursor, ciphertext_decrypted_b64);
+    fail_on(rc != AWS_OP_SUCCESS);
+    aws_byte_buf_append_null_terminator(ciphertext_decrypted_b64);
+
+    aws_nitro_enclaves_kms_client_destroy(client);
+    aws_credentials_release(credentials);
+    return;
+}
+
+int main(int argc, char **argv) {
+    struct app_ctx app_ctx;
```

Comment:
You can use `AWS_ZERO_STRUCT` here to clear the whole structure before setting any of its fields.

Reply:
The struct variables will be initialized in `s_parse_options()`